sdk/python/kfp/client/auth.py

```diff
@@ -179,40 +190,149 @@ def get_google_open_id_connect_token(service_account_credentials):
     return token_response['id_token']
 
 
-def get_refresh_token_from_client_id(client_id, client_secret):
+def get_refresh_token_from_client_id(client_id: str, client_secret: str):
     """Obtains the ID token for provided Client ID with user accounts.
 
     Flow: get authorization code -> exchange for refresh token -> obtain and
     return ID token.
+
+    Args:
+        client_id (str): OAuth client ID
+        client_secret (str): OAuth client secret
+          https://console.cloud.google.com/apis/credentials
+    Returns:
+        str: OAuth short-lived access token or long-lived refresh token
+    """
+    auth_code, redirect_uri = get_auth_code(client_id)
+    token = get_refresh_token_from_code(auth_code, client_id, client_secret,
+                                        redirect_uri)
+    return token
+
+
+def get_auth_code(client_id: str):
+    """Retrieves authorization token using Loopback flow.
+
+    Args:
+        client_id (str): OAuth client ID from
+          https://console.cloud.google.com/apis/credentials
+    Returns:
+        str: authorization token
+        str: redirect_uri parameter
     """
-    auth_code = get_auth_code(client_id)
-    return get_refresh_token_from_code(auth_code, client_id, client_secret)
+    host = 'localhost'
+    port = 9999
```

Comment:
As users will likely invoke functions in `auth.py` from `client.py` I don't want to expose `port` to avoid confusion. For instance, `kfp.Client` constructor takes KF endpoint as `host` argument. Adding `local_port` or alike might confuse users. Nevertheless, `port` and `host` are defined only once (no hard-coded values in other places), which makes it easier to change in the future.

Another option that I think we should avoid is to check what ports are available and pick one. The reason is because we might need to change this flow in the future as it seems that Loopback flow has been already deprecated for Android/iOS. So, whatever we decide now is a temporary fix anyways.

Instead, I propose to fall back to manual copy-pasting of the response page's url. This should work as long as there is no redirect on `http://localhost:9901` that removes `code` parameter from the url. If there is a redirect, we can instruct users to temporarily disable a process that takes port 9901. I changed the port number from 9999 as I had no matches when searching for "localhost:9901" on GitHub. We can also set the default port number to something else between 1024 and 65535.

Let me know what you think.

Reply:
That makes sense in terms of falling back to manual copy-paste when the port is not available, because the Loopback flow is deprecated for Android/iOS. Thank you for the detailed explanation!